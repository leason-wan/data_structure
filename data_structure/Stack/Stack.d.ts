import { SinglyLinked } from "../Linked/Linked.d";

export interface Stack {
  head: SinglyLinked
}

export interface StackApis {
  top(): void
  pop(): void
  push(data: any): void
}