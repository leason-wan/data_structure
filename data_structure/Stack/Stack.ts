import { SinglyLinked } from '../Linked/Linked';
import { Stack, StackApis } from './Stack.d';

export function Stack(): StackApis {
  const state: Stack = {
    head: SinglyLinked(null)
  }
  const top = () => state.head.value;
  const pop = () => {
    state.head.next && (state.head = state.head.next);
  };
  const push = (data: any) => {
    const linked = SinglyLinked(data);
    linked.next = state.head;
    state.head = linked;
  };

  return {
    top,
    pop,
    push,
  }
}
