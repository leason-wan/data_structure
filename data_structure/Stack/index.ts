import { Stack } from './Stack';
import { StackApis } from './Stack.d';

const s: StackApis = Stack();
s.push(0);
s.push(1);
s.push(2);
console.log(s.top());
s.pop();
s.pop();
s.pop();
console.log(s.top());
