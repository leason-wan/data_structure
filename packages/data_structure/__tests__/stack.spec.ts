import Stack from '../src/stack'

describe('test stack', () => {
  const s = Stack();

  test('create stack', () => {
    expect(s).toMatchSnapshot();
  })

  test('push stack & toString', () => {
    s.push(0);
    s.push(1);
    s.push(2);
    expect(s.toString()).toBe('0 -> 1 -> 2');
  })

  test('top stack', () => {
    expect(s.top()).toBe(2);
  })

  test('pop stack', () => {
    expect(s.pop()).toBe(2);
  })

})