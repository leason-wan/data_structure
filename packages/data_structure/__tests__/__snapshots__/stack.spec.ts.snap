// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test stack create stack 1`] = `
Object {
  "pop": [Function],
  "push": [Function],
  "toString": [Function],
  "top": [Function],
}
`;
